Make l1_loss composite

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #79381
* #77975
* __->__ #78257

Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc @jbschlosser

lezcano committed:
Make l1_loss composite

Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc @jbschlosser

[ghstack-poisoned]

jbschlosser commented:
> As a side note, `l1_loss_out` was incorrect in a number of ways. Even more, it is not exposed to the public as `F.l1_loss` does not accept an `out=` parameter. As such it is not even tested. I wonder how useful is to have `out=` variants for loss functions if we don't expose them at all. Even more, I wonder how useful is to have `_out` variants for loss functions, given that their most normal use case is to return just a real number cc @jbschlosser

Right, we don't expose out variants to Python for any of the loss functions in `torch.nn.functional`. Historically (and I think this is still the case), out variants and autograd didn't play well together so it didn't make any sense to have out variants for loss functions, as their core purpose is for training. Agreed it also makes little sense for `reduction != 'none'`, and I'm not sure why these out variants were introduced in the first place. Looking through the history, it seems they were ported to ATen from TH like this - maybe there was some TH-specific reason for these existing, idk. I certainly wouldn't add more out variants for losses, and maybe we can get rid of them eventually, but let's keep `l1_loss_out` in for this PR.

Lezcano commented:
This PR would need a matching PR in XLA @JackCaoG

Lezcano commented:
The benchmarks for forward + backward:
Faster in GPU, slower in CPU (interesting). I run them twice, and they seem to be consistent. My guess would be that, on CUDA, the composite kernel is able to save some intermediate computations (the `self - target` one, for example), so it's able to launch less kernels. On the other hand, note that the new implementation is more correct (gives the correct subgradient of zero at zero) but it needs to call `masked_fill`. This one may be slow on CPU (???). I really don't know about the CPU part really...
```
[----------- L1 loss cpu -----------]
                            |  master
4 threads: --------------------------
      size: (1000, 1000)    |     376
      size: (10000, 10000)  |  331000

Times are in microseconds (us).

[----------- L1 loss cuda ----------]
                            |  master
4 threads: --------------------------
      size: (1000, 1000)    |    407
      size: (10000, 10000)  |  37650

Times are in microseconds (us).

[----------- L1 loss cpu ------------]
                            |  This PR
4 threads: ---------------------------
      size: (1000, 1000)    |      580
      size: (10000, 10000)  |   434000

Times are in microseconds (us).

[----------- L1 loss cuda -----------]
                            |  This PR
4 threads: ---------------------------
      size: (1000, 1000)    |     293
      size: (10000, 10000)  |   26870

Times are in microseconds (us).
```

The script:
```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, device):
    a = torch.rand(*size, device=device, requires_grad=True)
    b = torch.rand(*size, device=device, requires_grad=True)
    print(a.shape, device)

    timer = Timer(
        "torch.nn.functional.l1_loss(a, b).backward()",
        globals={"a": a, "b": b},
        label=f"L1 loss {device}",
        sub_label=f"size: {size}",
        description="master",
        num_threads=4)

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    sizes = ((1000,)*2, (10000,)*2)
    for size, device in product(sizes, ("cpu", "cuda")):
        yield size, device


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
```

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

JackCaoG commented:
@Lezcano Can you give me a ping when this pr is approved? XLA side fix should be relatively quick 😄

ngimel commented:
We should not have `_out` variants if they are not exposed (that is, it's ok to delete them).

ngimel commented:
@lezcano can you please collect cpu and gpu profiles (`with torch.profiler.profile() as p:`) to shed some more light on perf differences?

Lezcano commented:
So to be able to figure out what's going on here I split the benchmarks and profiling in three: forward (the operation itself), backward, and forward&backward.

In the benchmarks we see that the forward on CPU is slightly slower, while the backward is slightly faster. It's not clear to me why CPU is slightly faster. On forward it's clear that both implementations are equivalent. In the backward though, this PR is much better, as it does half the computations. In this PR the backward is computed once up to `x - t`, and then it sets `grad_t = -grad_x` (note the call to `neg`) otoh, in the current implementation (which happens for many loss functions with manually implemented backwards btw) this optimisation is not done, and rather the backward is computed yet again (see the number of calls to each function in the profiler).

So, it does make sense that this PR is better as it has a more efficient backward. Now the benchmarks (I rerun them all to be sure, and was 100% sure that there wasn't any other job in the box) do not seem to say this on CPU! For example, I do not understand why in the profiler the same call to `sub` in the forward is noticeably faster on `master` than on this PR! The backward on CPU is even worse. This PR clearly does less operations than `master`, but it is still slower!

As an extra thing that I don't understand, `abs` seems to be called twice both for this PR and `master`, but in the code we clearly just call it once...


<details>
<summary>
Benchmark
</summary>

```
[-------- L1 loss cpu -------]
               |  master | PR
4 threads: -------------------
      fwd      |    84  |  134
      bwd      |   244  |  300
      fwd+bwd  |   330  |  434

Times are in milliseconds (ms).

[-------- L1 loss cuda --------]
               |  master | PR
4 threads: ---------------------
      fwd      |   8.02 |   8.03
      bwd      |  29.63 |  18.84
      fwd+bwd  |  37.65 |  26.87

Times are in milliseconds (ms).

```

</details>

<details>
<summary>
Benchmark script
</summary>

```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, device, fn, name):
    a = torch.rand(*size, device=device, requires_grad=True)
    b = torch.rand(*size, device=device, requires_grad=True)
    z = torch.nn.functional.l1_loss(a, b)
    print(a.shape, device)

    timer = Timer(
        fn,
        globals={"a": a, "b": b, "z": z},
        label=f"L1 loss {device}",
        sub_label=name,
        description="master",
        num_threads=4)

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    fns = [("fwd", "torch.nn.functional.l1_loss(a, b)"),
           ("bwd", "z.backward(retain_graph=True)"),
           ("fwd+bwd", "torch.nn.functional.l1_loss(a, b).backward()")]
    for device, (name, fn) in product(("cpu", "cuda"), fns):
        yield (10000, 10000), device, fn, name


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
``` 

</details>


<details>
<summary>
This PR cpu
</summary>

```
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                       Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                  aten::sub        42.50%      32.977ms        42.50%      32.977ms      32.977ms             1
                  aten::abs        24.87%      19.301ms        49.67%      38.547ms      19.273ms             2
              aten::l1_loss        23.37%      18.137ms        99.99%      77.594ms      77.594ms             1
                  aten::sum         8.87%       6.885ms         8.94%       6.937ms       6.937ms             1
                 aten::mean         0.15%     114.000us         9.24%       7.173ms       7.173ms             1
                 aten::div_         0.09%      71.000us         0.16%     122.000us     122.000us             1
                aten::copy_         0.04%      32.000us         0.04%      32.000us      16.000us             2
                   aten::to         0.03%      24.000us         0.07%      51.000us      51.000us             1
           aten::as_strided         0.02%      17.000us         0.02%      17.000us       5.667us             3
             aten::_to_copy         0.02%      14.000us         0.03%      27.000us      27.000us             1
                aten::empty         0.01%       7.000us         0.01%       7.000us       3.500us             2
    aten::broadcast_tensors         0.01%       5.000us         0.01%       5.000us       5.000us             1
                aten::fill_         0.01%       5.000us         0.01%       5.000us       5.000us             1
               aten::select         0.01%       5.000us         0.01%      11.000us      11.000us             1
        aten::empty_strided         0.01%       4.000us         0.01%       4.000us       4.000us             1
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 77.599ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
      autograd::engine::evaluate_function: AbsBackward0        26.70%      35.937ms        73.69%      99.202ms      99.202ms             1
                                              aten::mul        18.29%      24.625ms        18.29%      24.625ms      24.625ms             1
                                              aten::sgn        15.12%      20.358ms        15.12%      20.358ms      20.358ms             1
                                              aten::neg        13.76%      18.517ms        13.76%      18.517ms      18.517ms             1
                                           AbsBackward0        13.58%      18.282ms        47.00%      63.265ms      63.265ms             1
                                              aten::div        12.39%      16.678ms        12.40%      16.695ms      16.695ms             1
                                           SubBackward0         0.03%      43.000us        13.79%      18.560ms      18.560ms             1
                        torch::autograd::AccumulateGrad         0.03%      42.000us         0.04%      57.000us      28.500us             2
                                          MeanBackward0         0.02%      31.000us        12.44%      16.750ms      16.750ms             1
                                           aten::expand         0.02%      21.000us         0.02%      24.000us      24.000us             1
      autograd::engine::evaluate_function: SubBackward0         0.01%      16.000us        13.80%      18.576ms      18.576ms             1
                                                 detach         0.01%      13.000us         0.01%      13.000us       6.500us             2
     autograd::engine::evaluate_function: MeanBackward0         0.01%      10.000us        12.45%      16.760ms      16.760ms             1
                                       aten::empty_like         0.01%       7.000us         0.01%      12.000us      12.000us             1
                                         aten::_to_copy         0.01%       7.000us         0.01%      15.000us      15.000us             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 134.619ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
      autograd::engine::evaluate_function: AbsBackward0        17.99%      36.318ms        49.07%      99.054ms      99.054ms             1
                                              aten::mul        12.05%      24.331ms        12.05%      24.331ms      24.331ms             1
                                              aten::sub        11.96%      24.146ms        11.96%      24.146ms      24.146ms             1
                                              aten::sgn         9.99%      20.167ms         9.99%      20.167ms      20.167ms             1
                                              aten::neg         9.19%      18.547ms         9.19%      18.547ms      18.547ms             1
                                              aten::abs         9.16%      18.498ms        18.32%      36.979ms      18.489ms             2
                                          aten::l1_loss         9.04%      18.241ms        33.46%      67.548ms      67.548ms             1
                                           AbsBackward0         9.04%      18.238ms        31.08%      62.736ms      62.736ms             1
                                              aten::div         8.22%      16.601ms         8.23%      16.618ms      16.618ms             1
                                              aten::sum         3.26%       6.580ms         3.27%       6.597ms       6.597ms             1
                                             aten::mean         0.01%      29.000us         3.30%       6.659ms       6.659ms             1
                                            aten::copy_         0.01%      19.000us         0.01%      19.000us       6.333us             3
                                             aten::div_         0.01%      14.000us         0.02%      33.000us      33.000us             1
                                         aten::_to_copy         0.01%      14.000us         0.02%      31.000us      15.500us             2
      autograd::engine::evaluate_function: SubBackward0         0.00%      10.000us         9.20%      18.567ms      18.567ms             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 201.859ms


```

</details>

<details>
<summary>
This PR CUDA
</summary>

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub         1.60%     134.000us         5.68%     475.000us     475.000us       3.952ms        49.96%       3.952ms       3.952ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.952ms        49.96%       3.952ms       3.952ms             1
                                              aten::abs         0.53%      44.000us         9.33%     780.000us     390.000us       2.690ms        34.01%       5.380ms       2.690ms             2
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.690ms        34.01%       2.690ms       2.690ms             1
                                             aten::mean         0.60%      50.000us         2.15%     180.000us     180.000us       1.268ms        16.03%       1.268ms       1.268ms             1
void at::native::reduce_kernel<512, 1, at::native::R...         0.00%       0.000us         0.00%       0.000us       0.000us       1.268ms        16.03%       1.268ms       1.268ms             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 8.363ms
Self CUDA time total: 7.910ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::mul         0.34%      37.000us         5.68%     619.000us     619.000us       3.953ms        37.24%       3.953ms       3.953ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.953ms        37.24%       3.953ms       3.953ms             1
                                              aten::sgn         0.33%      36.000us         7.24%     789.000us     789.000us       2.695ms        25.39%       2.695ms       2.695ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.695ms        25.39%       2.695ms       2.695ms             1
                                              aten::neg         0.24%      26.000us         0.37%      40.000us      40.000us       2.688ms        25.32%       2.688ms       2.688ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.688ms        25.32%       2.688ms       2.688ms             1
                                              aten::div         0.78%      85.000us         1.28%     139.000us     139.000us       1.278ms        12.04%       1.278ms       1.278ms             1
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us       1.278ms        12.04%       1.278ms       1.278ms             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 10.891ms
Self CUDA time total: 10.616ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub         0.79%     149.000us         2.82%     533.000us     533.000us       3.955ms        21.33%       3.955ms       3.955ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.955ms        21.33%       3.955ms       3.955ms             1
                                              aten::mul         0.11%      21.000us         3.14%     595.000us     595.000us       3.949ms        21.30%       3.949ms       3.949ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.949ms        21.30%       3.949ms       3.949ms             1
                                              aten::abs         0.06%      12.000us         3.95%     748.000us     374.000us       2.690ms        14.51%       5.380ms       2.690ms             2
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.690ms        14.51%       2.690ms       2.690ms             1
                                              aten::sgn         0.12%      22.000us         3.54%     670.000us     670.000us       2.687ms        14.49%       2.687ms       2.687ms             1
                                              aten::neg         0.07%      13.000us         0.12%      22.000us      22.000us       2.687ms        14.49%       2.687ms       2.687ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.687ms        14.49%       2.687ms       2.687ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.687ms        14.49%       2.687ms       2.687ms             1
                                              aten::div         0.15%      28.000us         0.26%      49.000us      49.000us       1.302ms         7.02%       1.302ms       1.302ms             1
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us       1.302ms         7.02%       1.302ms       1.302ms             1
                                             aten::mean         0.10%      19.000us         0.16%      30.000us      30.000us       1.269ms         6.84%       1.269ms       1.269ms             1
void at::native::reduce_kernel<512, 1, at::native::R...         0.00%       0.000us         0.00%       0.000us       0.000us       1.268ms         6.84%       1.268ms       1.268ms             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 18.920ms
Self CUDA time total: 18.541ms
```

</details>

<details>
<summary>
master cpu
</summary>

```
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                       Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                  aten::sub        41.48%      26.149ms        41.48%      26.149ms      26.149ms             1
              aten::l1_loss        29.31%      18.478ms       199.87%     125.992ms      62.996ms             2
                  aten::abs        16.87%      10.636ms        16.87%      10.636ms      10.636ms             1
                  aten::sum        11.93%       7.520ms        12.02%       7.575ms       7.575ms             1
                 aten::mean         0.12%      77.000us        12.30%       7.754ms       7.754ms             1
                 aten::div_         0.09%      59.000us         0.16%     102.000us     102.000us             1
                   aten::to         0.04%      26.000us         0.07%      43.000us      43.000us             1
                aten::copy_         0.03%      21.000us         0.03%      21.000us      10.500us             2
           aten::as_strided         0.03%      16.000us         0.03%      16.000us       5.333us             3
             aten::_to_copy         0.02%      13.000us         0.03%      17.000us      17.000us             1
                aten::empty         0.02%      10.000us         0.02%      10.000us       5.000us             2
                 aten::abs_         0.02%      10.000us        16.89%      10.646ms      10.646ms             1
    aten::broadcast_tensors         0.01%       6.000us         0.01%       6.000us       6.000us             1
                aten::fill_         0.01%       6.000us         0.01%       6.000us       6.000us             1
               aten::select         0.01%       6.000us         0.01%       7.000us       7.000us             1
---------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 63.038ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub        39.53%      49.503ms        39.53%      49.503ms      24.752ms             2
                                            aten::fill_        25.99%      32.541ms        25.99%      32.541ms      10.847ms             3
                                             aten::mul_        17.78%      22.264ms        17.78%      22.264ms      11.132ms             2
                                             aten::sgn_        16.38%      20.508ms        16.38%      20.508ms      10.254ms             2
                                 aten::l1_loss_backward         0.09%     107.000us       173.68%     217.493ms      54.373ms             4
                                        L1LossBackward0         0.05%      62.000us        99.93%     125.137ms     125.137ms             1
                                              aten::div         0.03%      37.000us         0.06%      75.000us      37.500us             2
                        torch::autograd::AccumulateGrad         0.03%      35.000us         0.04%      55.000us      27.500us             2
                                       aten::zeros_like         0.03%      32.000us        26.05%      32.618ms      16.309ms             2
                                            aten::empty         0.02%      31.000us         0.02%      31.000us      15.500us             2
                                                 detach         0.01%      17.000us         0.01%      17.000us       8.500us             2
                                         aten::_to_copy         0.01%      15.000us         0.02%      31.000us      15.500us             2
                                       aten::empty_like         0.01%      14.000us         0.04%      51.000us      17.000us             3
autograd::engine::evaluate_function: L1LossBackward0...         0.01%      12.000us        99.94%     125.149ms     125.149ms             1
                                            aten::copy_         0.01%      12.000us         0.01%      12.000us       6.000us             2
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 125.225ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub        39.67%      73.390ms        39.67%      73.390ms      24.463ms             3
                                            aten::fill_        17.59%      32.549ms        17.59%      32.549ms       8.137ms             4
                                             aten::mul_        12.29%      22.738ms        12.29%      22.738ms      11.369ms             2
                                             aten::sgn_        10.84%      20.059ms        10.84%      20.059ms      10.030ms             2
                                          aten::l1_loss         9.91%      18.333ms        65.45%     121.075ms      60.538ms             2
                                              aten::abs         5.39%       9.966ms         5.39%       9.966ms       9.966ms             1
                                              aten::sum         4.09%       7.558ms         4.10%       7.584ms       7.584ms             1
                                 aten::l1_loss_backward         0.04%      79.000us       116.80%     216.065ms      54.016ms             4
                                            aten::empty         0.04%      73.000us         0.04%      73.000us      18.250us             4
                                              aten::div         0.02%      37.000us         0.04%      72.000us      36.000us             2
                                         aten::_to_copy         0.01%      23.000us         0.02%      45.000us      15.000us             3
                                            aten::copy_         0.01%      21.000us         0.01%      21.000us       5.250us             4
                                             aten::mean         0.01%      17.000us         4.13%       7.634ms       7.634ms             1
                                        L1LossBackward0         0.01%      16.000us        67.24%     124.384ms     124.384ms             1
                                             aten::div_         0.01%      15.000us         0.02%      33.000us      33.000us             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 184.994ms
```

</details>

<details>
<summary>
master CUDA
</summary>

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub         0.67%      57.000us         4.95%     419.000us     419.000us       3.951ms        49.96%       3.951ms       3.951ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.951ms        49.96%       3.951ms       3.951ms             1
                                              aten::abs         0.15%      13.000us         0.22%      19.000us      19.000us       2.688ms        33.99%       2.688ms       2.688ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.688ms        33.99%       2.688ms       2.688ms             1
                                             aten::mean         0.79%      67.000us         2.46%     208.000us     208.000us       1.269ms        16.05%       1.269ms       1.269ms             1
void at::native::reduce_kernel<512, 1, at::native::R...         0.00%       0.000us         0.00%       0.000us       0.000us       1.268ms        16.03%       1.268ms       1.268ms             1
                                             cudaMalloc         5.45%     461.000us         5.45%     461.000us     230.500us       0.000us         0.00%       0.000us       0.000us             2
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 8.464ms
Self CUDA time total: 7.908ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub         0.14%      32.000us         0.21%      47.000us      23.500us       7.897ms        36.89%       7.897ms       3.949ms             2
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       7.897ms        36.89%       7.897ms       3.949ms             2
                                             aten::mul_         0.23%      51.000us         0.29%      65.000us      32.500us       5.534ms        25.85%       5.534ms       2.767ms             2
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us       5.534ms        25.85%       5.534ms       2.767ms             2
                                             aten::sgn_         0.12%      27.000us         0.19%      42.000us      21.000us       5.371ms        25.09%       5.371ms       2.686ms             2
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       5.371ms        25.09%       5.371ms       2.686ms             2
                                            aten::fill_         0.22%      49.000us         0.57%     126.000us      42.000us       2.600ms        12.15%       2.600ms     866.667us             3
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.600ms        12.15%       2.600ms     866.667us             3
                                    aten::empty_strided         2.60%     574.000us         2.60%     574.000us     574.000us       0.000us         0.00%       0.000us       0.000us             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 22.118ms
Self CUDA time total: 21.407ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                              aten::sub         0.11%      33.000us         3.39%       1.042ms     347.333us      11.846ms        40.43%      11.846ms       3.949ms             3
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us      11.846ms        40.43%      11.846ms       3.949ms             3
                                             aten::mul_         0.07%      20.000us         0.10%      32.000us      16.000us       5.533ms        18.88%       5.533ms       2.767ms             2
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us       5.533ms        18.88%       5.533ms       2.767ms             2
                                             aten::sgn_         0.05%      16.000us         0.09%      28.000us      14.000us       5.370ms        18.33%       5.370ms       2.685ms             2
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       5.370ms        18.33%       5.370ms       2.685ms             2
                                              aten::abs         0.02%       5.000us         0.04%      11.000us      11.000us       2.688ms         9.17%       2.688ms       2.688ms             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.688ms         9.17%       2.688ms       2.688ms             1
                                            aten::fill_         0.08%      25.000us         0.22%      69.000us      23.000us       2.593ms         8.85%       2.593ms     864.333us             3
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.593ms         8.85%       2.593ms     864.333us             3
                                             aten::mean         0.06%      18.000us         0.10%      31.000us      31.000us       1.268ms         4.33%       1.268ms       1.268ms             1
void at::native::reduce_kernel<512, 1, at::native::R...         0.00%       0.000us         0.00%       0.000us       0.000us       1.267ms         4.32%       1.267ms       1.267ms             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 30.750ms
Self CUDA time total: 29.302ms
```

</details>


<details>
<summary>
Profiling script
</summary>
```python
import torch


def profile(device):
    act = dict(cpu=[torch.profiler.ProfilerActivity.CPU],
               cuda=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA])[device]
    sort_by = f"self_{device}_time_total"

    x = torch.rand((10000, 10000), device=device, requires_grad=True)
    y = torch.rand((10000, 10000), device=device, requires_grad=True)
    if device == "cuda":
        torch.cuda.synchronize()
    with torch.profiler.profile(activities=act) as p:
        torch.nn.functional.l1_loss(x, y)

    print(p.key_averages().table(sort_by=sort_by, row_limit=15))

    x = torch.rand((10000, 10000), device=device, requires_grad=True)
    y = torch.rand((10000, 10000), device=device, requires_grad=True)
    z = torch.nn.functional.l1_loss(x, y)
    if device == "cuda":
        torch.cuda.synchronize()
    with torch.profiler.profile(activities=act) as p:
        z.backward()

    print(p.key_averages().table(sort_by=sort_by, row_limit=15))

    x = torch.rand((10000, 10000), device=device, requires_grad=True)
    y = torch.rand((10000, 10000), device=device, requires_grad=True)
    if device == "cuda":
        torch.cuda.synchronize()
    with torch.profiler.profile(activities=act) as p:
        torch.nn.functional.l1_loss(x, y).backward()

    print(p.key_averages().table(sort_by=sort_by, row_limit=15))


profile("cpu")
profile("cuda")
```

</details>

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

mruberry commented:
Offline feedback: 

- try averaging profiles over many iterations
- this is mergeable without addressing CPU issue
- follow-up on CPU performance discrepancy -- seems like something interesting is happening

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

jbschlosser commented:
@Lezcano Thanks for looking into it. I'm good with getting this in anyway and opening a new issue for the CPU perf. CUDA perf is vastly more important anyway.

Lezcano commented:
@jbschlosser can I have a LGTM then?
@JackCaoG this is ready to get the fix from the XLA side.

jbschlosser reviewed (approved):
LGTM :)

Lezcano commented:
So, according to my benchmarks, `abs_()` is between x3 - x4 faster than `abs()` on CPU on a float32 tensor of 10^8 elements. That's where the CPU efficiency loss in this PR is going.

According to the profiler, almost all this time is spent in the TI (construction and execution). I have not dug any further. Do you think this is worth following up in an issue  @ngimel ?

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

JackCaoG commented:
@Lezcano Can we rebase, update xla pin and merge this pr?

lezcano committed:
Update on "Make l1_loss composite"


Fixing the forward AD for `sgn` in the next PR of this stack uncovered a
number of issues with the derivatives of `l1_loss`. Upon inspection,
`l1_loss` was just implemented as a composite function, but it was not
differentiable. This PR makes it a fully differentiable function.

As a side note, `l1_loss_out` was incorrect in a number of ways. Even
more, it is not exposed to the public as `F.l1_loss` does not accept an
`out=` parameter. As such it is not even tested. I wonder how useful is
to have `out=` variants for loss functions if we don't expose them at
all. Even more, I wonder how useful is to have `_out` variants  for loss
functions, given that their most normal use case is to return just a
real number cc jbschlosser

[ghstack-poisoned]

Lezcano commented:
I just did. If it looks good could you merge it? @JackCaoG

JackCaoG commented:
oh I don't have the merge access lol. Maybe @bdhirsh or @ngimel can help? I am not sure if bot can merge this since it is already approved.

Lezcano commented:
Thank you so much @ngimel!

Lezcano commented:
Fwiw, I changed `.github/ci_commit_pins/xla.txt` even though @JackCaoG in https://github.com/pytorch/xla/pull/3624#issuecomment-1156848175 mentioned the file `.github/xla_commit_hash.txt` as that file does not exist in master. Now, the hud https://hud2.pytorch.org/ says that my PR broke some tests so I guess it's going to be reverted.

So... do I changed the correct file with the correct value? I guess I didn't but I don't know what I did wrong @bdhirsh

JackCaoG commented:
I think you pin to the right pr.. but for some reason test started to fail on xla end as well. Let's revert, I will invesgate where the new failure is coming from. It might or might not be this pr.

JackCaoG commented:
I started to see
```
======================================================================
FAIL: test_nn_scalars_reductions_xla (__main__.TestNNDeviceTypeXLA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/tmp/pytorch/xla/test/../../test/test_nn.py", line 15479, in test_nn_scalars_reductions
    verify_reduction_scalars(input, reduction, output)
  File "/tmp/pytorch/xla/test/../../test/test_nn.py", line 15462, in verify_reduction_scalars
    self.assertEqual((), output.shape)
  File "/tmp/pytorch/xla/test/pytorch_test_base.py", line 628, in assertEqual
    return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2273, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1068, in assert_equal
    raise error_meta.to_error() from None
AssertionError: The length of the sequences mismatch: 0 != 2

----------------------------------------------------------------------
```
on my pytorch/xla master. I will look into that..

JackCaoG commented:
weird.. I was not able to repo this issue locally with pytorch and pytorch/xla head..

Lezcano commented:
So, I think it was a bug in CI. See that the PR in the hud after mine is this revert https://github.com/pytorch/pytorch/pull/79470 which is relevant and makes the failing CI job go green again. So perhaps this PR is fine? cc @suo @janeyx99

JackCaoG commented:
yea.. I think I will trigger the CI again. We should be good.

malfet commented:
This breaks internal build (there is another project that depends on old definition of `l1_loss`, trying to forward fix it now

malfet commented:
@pytorchbot revert -m "This breaks executorch" -c weird

zou3519 commented:
The current policy for breaking the internal (fb) version of functorch is to revert the PR to prevent downstream internal breakages. This is unfortunately disruptive to developers (sorry for the inconvenience, @Lezcano), but is a temporary measure and will be resolved once we execute on https://github.com/pytorch/pytorch/issues/78082 after the 1.12 release.

@Lezcano if you could please re-submit this PR, as a workaround, we (someone from the functorch team) can help land it via the old import workflow rather than through the github-first workflow.
